File: llm_written_v1/gemini-2.5-pro/smith-waterman/java/SmithWatermanParallel.java
import java.util.concurrent.ForkJoinPool;
import java.util.stream.IntStream;

public class SmithWatermanParallel {
    private final int matchScore;
    private final int mismatchScore;
    private final int gapScore;
    private final ForkJoinPool forkJoinPool;
    // Increased threshold: parallelism is only beneficial for very large matrices.
    private static final int SEQUENTIAL_THRESHOLD = 2000 * 2000; 

    public SmithWatermanParallel(int matchScore, int mismatchScore, int gapScore) {
        this.matchScore = matchScore;
        this.mismatchScore = mismatchScore;
        this.gapScore = gapScore;
        // Bounded to the number of available processors
        this.forkJoinPool = new ForkJoinPool(Runtime.getRuntime().availableProcessors());
    }

    public int[][] constructMatrix(String query, String reference) {
        int n = query.length() + 1;
        int m = reference.length() + 1;
        int[][] H = new int[n][m];

        // Fallback for small inputs
        if ((long) n * m < SEQUENTIAL_THRESHOLD) {
            return constructMatrixSequential(query, reference, H);
        }

        // The number of anti-diagonals is n + m - 1
        for (int k = 1; k < n + m - 1; k++) {
            final int current_k = k;
            // The range of i for a given anti-diagonal k
            int i_start = Math.max(1, current_k - m + 1);
            int i_end = Math.min(n, current_k + 1);

            // Execute the calculation for the current anti-diagonal in parallel
            forkJoinPool.submit(() -> {
                IntStream.range(i_start, i_end).parallel().forEach(i -> {
                    int j = current_k - i + 1;
                    if (j > 0 && j < m) {
                        int scoreDiagonal = H[i - 1][j - 1] +
                                (query.charAt(i - 1) == reference.charAt(j - 1) ? matchScore : mismatchScore);
                        int scoreUp = H[i - 1][j] + gapScore;
                        int scoreLeft = H[i][j - 1] + gapScore;
                        H[i][j] = Math.max(0, Math.max(scoreDiagonal, Math.max(scoreUp, scoreLeft)));
                    }
                });
            }).join(); // Wait for the current anti-diagonal to complete before starting the next
        }

        return H;
    }

    private int[][] constructMatrixSequential(String query, String reference, int[][] H) {
        int n = H.length;
        int m = H[0].length;
        for (int i = 1; i < n; i++) {
            for (int j = 1; j < m; j++) {
                int scoreDiagonal = H[i - 1][j - 1] +
                        (query.charAt(i - 1) == reference.charAt(j - 1) ? matchScore : mismatchScore);
                int scoreUp = H[i - 1][j] + gapScore;
                int scoreLeft = H[i][j - 1] + gapScore;
                H[i][j] = Math.max(0, Math.max(scoreDiagonal, Math.max(scoreUp, scoreLeft)));
            }
        }
        return H;
    }


    public int[] findHighestScore(int[][] H) {
        int maxScore = 0;
        int maxI = 0;
        int maxJ = 0;
        for (int i = 0; i < H.length; i++) {
            for (int j = 0; j < H[i].length; j++) {
                if (H[i][j] > maxScore) {
                    maxScore = H[i][j];
                    maxI = i;
                    maxJ = j;
                }
            }
        }
        return new int[]{maxI, maxJ};
    }

    public AlignmentResult traceback(int[][] H, String query, String reference) {
        StringBuilder alignedA = new StringBuilder();
        StringBuilder alignedB = new StringBuilder();
        int[] maxPos = findHighestScore(H);
        int i = maxPos[0];
        int j = maxPos[1];
        int score = H[i][j];
        int totalMatch = 0;
        int totalAlignment = 0;

        while (i > 0 && j > 0) {
            int currentScore = H[i][j];
            if (currentScore == 0) break;

            int diagonalScore = H[i - 1][j - 1];
            int upScore = H[i - 1][j];
            int leftScore = H[i][j - 1];
            int expectedDiagonal = diagonalScore +
                    (query.charAt(i - 1) == reference.charAt(j - 1) ? matchScore : mismatchScore);

            if (currentScore == expectedDiagonal) {
                alignedA.append(query.charAt(i - 1));
                alignedB.append(reference.charAt(j - 1));
                totalAlignment++;
                if (query.charAt(i - 1) == reference.charAt(j - 1)) {
                    totalMatch++;
                }
                i--;
                j--;
            } else if (currentScore == upScore + gapScore) {
                alignedA.append(query.charAt(i - 1));
                alignedB.append('-');
                totalAlignment++;
                i--;
            } else if (currentScore == leftScore + gapScore) {
                alignedA.append('-');
                alignedB.append(reference.charAt(j - 1));
                totalAlignment++;
                j--;
            } else {
                break;
            }
        }

        alignedA.reverse();
        alignedB.reverse();
        double percentageIdentity = (totalAlignment > 0) ? ((double) totalMatch / totalAlignment) * 100.0 : 0.0;
        return new AlignmentResult(alignedA.toString(), alignedB.toString(), score, percentageIdentity);
    }

    public AlignmentResult findAlignment(String query, String reference) {
        int[][] H = constructMatrix(query, reference);
        return traceback(H, query, reference);
    }
    
    public void shutdown() {
        forkJoinPool.shutdown();
    }

    public static class AlignmentResult {
        public final String alignedA;
        public final String alignedB;
        public final int score;
        public final double identity;

        public AlignmentResult(String alignedA, String alignedB, int score, double identity) {
            this.alignedA = alignedA;
            this.alignedB = alignedB;
            this.score = score;
            this.identity = identity;
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) return true;
            if (obj == null || getClass() != obj.getClass()) return false;
            AlignmentResult that = (AlignmentResult) obj;
            return score == that.score &&
                   Double.compare(that.identity, identity) == 0 &&
                   alignedA.equals(that.alignedA) &&
                   alignedB.equals(that.alignedB);
        }

        @Override
        public String toString() {
            return "AlignmentResult{" +
                   "alignedA='" + alignedA + '\'' +
                   ", alignedB='" + alignedB + '\'' +
                   ", score=" + score +
                   ", identity=" + identity +
                   '}';
        }
    }
}

File: llm_written/gemini-2-5-pro/smith-waterman/cpp/smith_waterman_parallel.cpp
#include "smith_waterman_parallel.h"
#include "smith_waterman_sequential.h"
#include <algorithm>
#include <vector>
#include <omp.h>

SmithWaterman::SmithWaterman(int match, int mismatch, int gap) 
    : matchScore(match), mismatchScore(mismatch), gapScore(gap) {}

std::vector<std::vector<int>> SmithWaterman::constructMatrix(
    const std::string& query, 
    const std::string& reference) {
    
    int n = query.length() + 1;
    int m = reference.length() + 1;
    
    std::vector<std::vector<int>> H(n, std::vector<int>(m, 0));

    if (n <= 1 || m <= 1) {
        return H;
    }

    // Parallelize along anti-diagonals
    for (int k = 2; k <= n + m - 2; ++k) {
        #pragma omp parallel for schedule(static)
        for (int i = std::max(1, k - m + 1); i <= std::min(n - 1, k - 1); ++i) {
            int j = k - i;
            int scoreDiagonal = H[i-1][j-1] + 
                (query[i-1] == reference[j-1] ? matchScore : mismatchScore);
            
            int scoreUp = H[i-1][j] + gapScore;
            int scoreLeft = H[i][j-1] + gapScore;
            
            H[i][j] = std::max({0, scoreDiagonal, scoreUp, scoreLeft});
        }
    }
    
    return H;
}

std::pair<int, int> SmithWaterman::findHighestScore(
    const std::vector<std::vector<int>>& H) {
    
    int maxScore = 0;
    std::pair<int, int> maxPos = {0, 0};
    int n = H.size();
    if (n == 0) return maxPos;
    int m = H[0].size();

    #pragma omp parallel
    {
        int localMaxScore = 0;
        std::pair<int, int> localMaxPos = {0, 0};

        #pragma omp for collapse(2) schedule(static)
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (H[i][j] > localMaxScore) {
                    localMaxScore = H[i][j];
                    localMaxPos = {i, j};
                }
            }
        }

        #pragma omp critical
        {
            if (localMaxScore > maxScore) {
                maxScore = localMaxScore;
                maxPos = localMaxPos;
            } else if (localMaxScore == maxScore) {
                // Deterministic tie-breaking: choose the one with the smaller row, then column index
                if (localMaxPos.first < maxPos.first || 
                   (localMaxPos.first == maxPos.first && localMaxPos.second < maxPos.second)) {
                    maxPos = localMaxPos;
                }
            }
        }
    }
    
    return maxPos;
}

std::tuple<std::string, std::string, int, double> SmithWaterman::traceback(
    const std::vector<std::vector<int>>& H,
    const std::string& query,
    const std::string& reference) {
    
    std::string alignedA = "";
    std::string alignedB = "";
    
    auto [i, j] = findHighestScore(H);
    int score = H[i][j];
    
    int totalMatch = 0;
    int totalAlignment = 0;
    
    while (i > 0 && j > 0) {
        int currentScore = H[i][j];
        
        if (currentScore == 0) {
            break;
        }
        
        int diagonalScore = H[i-1][j-1];
        int upScore = H[i-1][j];
        int leftScore = H[i][j-1];
        
        int expectedDiagonal = diagonalScore + 
            (query[i-1] == reference[j-1] ? matchScore : mismatchScore);
        
        if (currentScore == expectedDiagonal) {
            alignedA += query[i-1];
            alignedB += reference[j-1];
            totalAlignment++;
            
            if (query[i-1] == reference[j-1]) {
                totalMatch++;
            }
            i--;
            j--;
        }
        else if (currentScore == upScore + gapScore) {
            alignedA += query[i-1];
            alignedB += '-';
            totalAlignment++;
            i--;
        }
        else if (currentScore == leftScore + gapScore) {
            alignedA += '-';
            alignedB += reference[j-1];
            totalAlignment++;
            j--;
        }
        else {
            // This case can happen if multiple paths give the same score.
            // To ensure determinism, we prioritize diagonal, then up, then left.
            alignedA += query[i-1];
            alignedB += reference[j-1];
            totalAlignment++;
            if (query[i-1] == reference[j-1]) {
                totalMatch++;
            }
            i--;
            j--;
        }
    }
    
    std::reverse(alignedA.begin(), alignedA.end());
    std::reverse(alignedB.begin(), alignedB.end());
    
    double percentageIdentity = (totalAlignment > 0) 
        ? (static_cast<double>(totalMatch) / totalAlignment) * 100.0 
        : 0.0;
    
    return {alignedA, alignedB, score, percentageIdentity};
}

std::tuple<std::string, std::string, int, double> SmithWaterman::findAlignment(
    const std::string& query,
    const std::string& reference) {
    
    if (query.length() < 500 || reference.length() < 500) {
        SmithWatermanSequential sw_seq(matchScore, mismatchScore, gapScore);
        return sw_seq.findAlignment(query, reference);
    }

    auto H = constructMatrix(query, reference);
    return traceback(H, query, reference);
}
